server/impl/nonce/nonce.go

```diff
@@ -0,0 +1,76 @@
+package nonce
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+)
+
+const (
+	stateLen       = 16
+	stateExpiry    = 30 * time.Minute
+	clearThreshold = 128
+)
+
+// Generator creates randomly generated 16-byte ASCII nonces. The nonce only has
+// Base32 characters, which are a subset of all alphanumeric characters.
+// Each nonce is only valid for 30 minutes; beyond that, and VerifyState will
+// return false.
+type Generator struct {
+	rand        *rand.Rand
+	m           map[string]time.Time
+	mu          sync.Mutex
+	lastCleaned int
+}
+
+// NewGenerator creates a new Generator from the given rand.Source.
+func NewGenerator(src rand.Source) *Generator {
+	return &Generator{
+		rand: rand.New(src),
+		m:    make(map[string]time.Time),
+	}
+}
+
+func (g *Generator) clean() {
+	if len(g.m)-g.lastCleaned >= clearThreshold {
+		for k, expiry := range g.m {
+			if time.Now().After(expiry) {
+				delete(g.m, k)
+			}
+		}
+		g.lastCleaned = len(g.m)
+	}
+}
+
+// NewState creates a new nonce and remembers it for later retrieval.
+func (g *Generator) NewState() string {
+	for {
+		state := RandomAlphanumerics(g.rand, stateLen)
```

Comment:
Apparently we need to move this inside the lock, since rand.Source is not concurrency-safe